spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/CassandraCatalog.scala

```diff
@@ -1,65 +1,157 @@
+    cachedDataSourceTables.invalidateAll()
+  }
 
   override def tableExists(tableIdentifier: Seq[String]): Boolean = {
     val (cluster, database, table) = getClusterDBTableNames(tableIdentifier)
-    val schema = schemas.get(cluster)
-    val tabDef = for (ksDef <- schema.keyspaceByName.get(database); tabDef <- ksDef.tableByName.get(table)) yield tabDef
-    tabDef.isDefined
+    val tableRef = TableRef(table, database, Option(cluster))
+    val tableIdent = fullTableIdent(tableIdentifier)
+    val cached = cachedDataSourceTables.asMap().containsKey(tableIdent)
+    if (cached) {
+      true
+    } else {
+      val schema = Schema.fromCassandra(getCassandraConnector(tableRef))
+      val tabDef =
+        for (ksDef <- schema.keyspaceByName.get(database);
+             tabDef <- ksDef.tableByName.get(table)) yield tabDef
+      tabDef.nonEmpty
+    }
+  }
+
+  override def getTables(databaseName: Option[String]): Seq[(String, Boolean)] = {
+    val cluster = csc.getCluster
+    val tableNamesFromCache = getTablesFromCache(databaseName, Option(cluster)).map(_._1)
+    val tablesFromCassandra = getTablesFromCassandra(databaseName)
+    val tablesOnlyInCache =
+      tableNamesFromCache.diff(tablesFromCassandra.map(_._1)).map(name => (name, true))
+
+    tablesFromCassandra ++ tablesOnlyInCache
+  }
+
+  /** List all tables for a given database name and cluster directly from Cassandra */
+  def getTablesFromCassandra(databaseName: Option[String]): Seq[(String, Boolean)] = {
+    val cluster = csc.getCluster
+    val tableRef = TableRef("", databaseName.getOrElse(""), Option(cluster))
+    val schema = Schema.fromCassandra(getCassandraConnector(tableRef))
+    if (databaseName.nonEmpty) {
+      val ksDef = schema.keyspaceByName.get(databaseName.get)
+      if (ksDef.nonEmpty) {
+        ksDef.get.tables.map(
+          tableDef => (s"${ksDef.get.keyspaceName}.${tableDef.tableName}", false)).toSeq
+      } else {
+        Seq.empty
+      }
+    } else {
+      val tables = mutable.Set[(String, Boolean)]()
+      for (ksDef <- schema.keyspaces) {
+        tables ++= ksDef.tables.map(
+          table => (s"${ksDef.keyspaceName}.${table.tableName}", false))
+      }
+      tables.toSeq
+    }
   }
 
-  override def getTables(databaseName: Option[String]): Seq[(String, Boolean)] = ???
+  /** List all tables for a given database name and cluster from local cache */
+  def getTablesFromCache(
+      databaseName: Option[String],
+      cluster: Option[String] = None): Seq[(String, Boolean)] = {
+    val iter = cachedDataSourceTables.asMap().keySet().iterator()
+    val clusterName = cluster.getOrElse(csc.getCluster)
+    val names = ListBuffer[String]()
+    while (iter.hasNext) {
+      val ident = iter.next()
+      val c = ident(0)
+      val db = ident(1)
+      val table = ident(2)
+      if (clusterName == c) {
+        if (databaseName.nonEmpty && databaseName.get == db) {
+          names += table
```

Comment:
Why in this case it is not qualified by the database name? Is this distinction really needed?
The code for getTableForCassandra did not make this distinction and qualified all table names with a database name.